adapters/sigfox-vertx/src/main/java/org/eclipse/hono/adapter/sigfox/impl/SigfoxProtocolAdapter.java

```diff
@@ -0,0 +1,176 @@
+                .method(HttpMethod.GET)
+                .method(HttpMethod.POST)
+                .method(HttpMethod.PUT)
+                .handler(ctx -> dataHandler(ctx, this::uploadTelemetryMessage));
+
+        router.route("/data/event")
+                .method(HttpMethod.GET)
+                .method(HttpMethod.POST)
+                .method(HttpMethod.PUT)
+                .handler(ctx -> dataHandler(ctx, this::uploadEventMessage));
+
+        router.errorHandler(500, t -> {
+            LOG.warn("Unhandled exception", t);
+        });
+    }
+
+    protected void dataHandler(final RoutingContext ctx, final UploadHandler uploadHandler) {
+
+        if (!(ctx.user() instanceof Device)) {
+            LOG.warn("Not a device");
+            return;
+        }
+
+        final Device gatewayDevice = (Device) ctx.user();
+
+        final String tenant = gatewayDevice.getTenantId();
+
+        final String deviceId = ctx.queryParams().get("device");
+        final Buffer data = decode(ctx.queryParams().get("data"));
+
+        LOG.debug("{} handler - deviceId: {}, data: {}", ctx.request().method(), deviceId, data);
+
+        final String contentType = (data != null) ? CONTENT_TYPE_OCTET_STREAM
+                : EventConstants.CONTENT_TYPE_EMPTY_NOTIFICATION;
+
+        uploadHandler.upload(ctx, tenant, deviceId, data, contentType);
+    }
+
+    @Override
+    protected void customizeDownstreamMessage(final Message downstreamMessage, final RoutingContext ctx) {
+        super.customizeDownstreamMessage(downstreamMessage, ctx);
+
+        // pass along all query parameters that start with 'sigfox.'
+        // If a key has multiple values, then only one of them will be mapped.
+
+        for (final Map.Entry<String, String> entry : ctx.queryParams()) {
+            if (entry.getKey() == null || !entry.getKey().startsWith(SIGFOX_PROPERTY_PREFIX)) {
+                continue;
+            }
+            downstreamMessage.getApplicationProperties().getValue().put(entry.getKey(), entry.getValue());
+        }
+
+        // pass along all headers that start with 'sigfox-' and map the prefix to "sigfox."
+        // If a key has multiple values, then only one of them will be mapped.
+
+        for (final Map.Entry<String, String> entry : ctx.request().headers()) {
+            String key = entry.getKey();
+            if (key == null) {
+                continue;
+            }
+            key = key.toLowerCase();
+            if (!key.startsWith(SIGFOX_HEADER_PREFIX)) {
+                continue;
+            }
+            key = key.substring(SIGFOX_HEADER_PREFIX.length());
+            downstreamMessage.getApplicationProperties().getValue().put(SIGFOX_PROPERTY_PREFIX + key, entry.getValue());
+        }
+
+    }
+
+    private static Buffer decode(final String data) {
+        if (data == null) {
+            return Buffer.buffer();
+        }
+        return Buffer.buffer(BaseEncoding.base16().decode(data.toUpperCase()));
```

Comment:
This is nice :+1:  I didn't know that guava contains a hex decoder. We can then also use it in the LoRaWan adapter and get rid of `org.eclipse.hono.util.Strings.decodeHex` ...

Reply:
I will do that in another PR.

Reply:
I have already pushed that change ...

Reply:
Ok … just saw your commit :+1: